### PR TITLE
use-true-false: Add framework for supporting known boolean expressions.

### PR DESCRIPTION
**Pull request by jamestalmage**
Partial Fix: #79

I didn't hunt down a complete list of boolean expressions, but I have provided a framework that makes it easy to edit the list.

It uses `call-signature` (which is part of what `power-assert` uses to do some of it's magic).

### CODE REVIEW

**Comment by sindresorhus** [comment]
Awesome. My contribution to this, I took the time to compile a list of builtins returning a boolean:

```
isFinite()
isNaN()
Object.is()
Object.isExtensible()
Object.isFrozen()
Object.isSealed()
Boolean()
Number.isNaN()
Number.isFinite()
Number.isInteger()
Number.isSafeInteger()
String.prototype.includes()
String.prototype.endsWith()
String.prototype.startsWith()
RegExp.prototype.test()
Array.isArray()
Array.prototype.includes()
Array.prototype.every()
Array.prototype.some()
Int8Array.prototype.every()
Int8Array.prototype.includes()
Int8Array.prototype.some()
Uint8Array.prototype.every()
Uint8Array.prototype.includes()
Uint8Array.prototype.some()
Uint8ClampedArray.prototype.every()
Uint8ClampedArray.prototype.includes()
Uint8ClampedArray.prototype.some()
Int16Array.prototype.every()
Int16Array.prototype.includes()
Int16Array.prototype.some()
Uint16Array.prototype.every()
Uint16Array.prototype.includes()
Uint16Array.prototype.some()
Int32Array.prototype.every()
Int32Array.prototype.includes()
Int32Array.prototype.some()
Uint32Array.prototype.every()
Uint32Array.prototype.includes()
Uint32Array.prototype.some()
Float32Array.prototype.every()
Float32Array.prototype.includes()
Float32Array.prototype.some()
Float64Array.prototype.every()
Float64Array.prototype.includes()
Float64Array.prototype.some()
Map.prototype.has()
Set.prototype.has()
WeakMap.prototype.has()
WeakSet.prototype.has()
ArrayBuffer.isView()
SharedArrayBuffer.isView()
Reflect.has()
Reflect.isExtensible()
```

Should I put in a module? Would there be any value to that?

**Comment by jamestalmage** [comment]
> should I put it in a module? Would there be any value to that?

Maybe - who knows. Probably more so if it had the full method signature:

```
Array.isArray(obj)
Object.is(value1, value2)
obj.fn(arg1, arg2, [optional_arg])
```

Also, `call-signature` is not smart enough to parse `String.prototype.includes`. It only allows member expressions one deep. It was a simplification of [`escallmatch`](https://github.com/twada/escallmatch) that uses `regexp` to parse the signature to avoid depending on a full JS parser (`escallmatch` relies on esprima). We could just use `espree` to parse the matcher templates, that would give more robust parsing without adding another dependency.

Still, I am not sure how to handle stuff on the prototype. Were you hoping to try to implement static analysis and figure out if `x.some()` was being called on an Array?

**Comment by twada** [comment]
> We could just use espree to parse the matcher templates, that would give more robust parsing without adding another dependency.

:+1: on @jamestalmage

**Comment by jamestalmage** [comment]
updated to use espree. I didn't realize we already had `espurify` and `deep-strict-equal` as dependencies. Those make the matcher really easy.

I haven't added the full list of methods yet. Still waiting on an answer to this:

> Were you hoping to try to implement static analysis and figure out if x.some() was being called on an Array?

Question was directed at @sindresorhus - but feel free to chime in.

File: package.json [hunk 1]
```diff
@@ -61,7 +61,9 @@
     "mocha"
   ],
   "dependencies": {
+    "call-signature": "0.0.2",
```
Comment (jfmengels) [inline package.json]: Not needed anymore

**Comment by jfmengels** [comment]
This looks good! Nice work @jamestalmage!

**Comment by jamestalmage** [comment]
I went with just the non `prototype` methods, as doing the rest well would require an enormous amount of static type analysis/inferencing.

It would be cool if someone built a linter with a type analysis engine, but I think that's a bit too ambitious to achieve this one feature.

**Comment by sindresorhus** [comment]
> I went with just the non prototype methods, as doing the rest well would require an enormous amount of static type analysis/inferencing.

:+1: 

> It would be cool if someone built a linter with a type analysis engine, but I think that's a bit too ambitious to achieve this one feature.

That is my dream. So many of my rule suggestions on ESLint has been turned down for requiring type analysis... Linting could be so much better with best effort type inference. [Flow](https://github.com/facebook/flow) compiled down to asm.js anyone??

**Comment by sindresorhus** [comment]
Can you add an example of `Array.isArray()` to its docs?

**Comment by sindresorhus** [comment]
Looks good to me. @twada @jfmengels Can you give it a final review?

**Comment by jfmengels** [comment]
LGTM :)

**Comment by jamestalmage** [comment]
> Flow compiled down to asm.js anyone

Even without Flow, Webstorm does an impressive job of type inference from pure JS code. I am sure that was no small undertaking on IntelliJ's part, but I think there are a number of portential uses for an open source alternative. (Atom would surely benefit)

**Comment by sindresorhus** [comment]
@jamestalmage Flow does that too. Its type notation is optional and only there to improve the inference.

There's also https://github.com/ternjs/tern

**Comment by twada** [comment]
LGTM :+1:

**Comment by sindresorhus** [comment]
Awesome! This makes the rule really solid.

**Comment by jamestalmage** [comment]
> Flow does that too. Its type notation is optional and only there to improve the inference

I guess I knew that. I haven't played with Flow much. Does it automatically analyze exports of imported files? It looks line `tern` does, though it appears to require launching a server.

**Pull request closed without merging**